### PR TITLE
Config refactoring

### PR DESCRIPTION
**Pull request by BlackSamorez**
Right now configs include:
* Plain text descriptions ("slice", "scale", etc.)
* lambdas
* partials
* function pointers

It makes it hard to read. I'm planning to standardize how it should look

### CODE REVIEW

File: src/tensor_parallel/autoconfig.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import torch
+import torch.distributed
+from torch import nn
+from torch.nn.modules import conv
+
+from tensor_parallel.slicer_wrapper import Config
+from tensor_parallel.state_actions import Scale, Split, SplitInsideChunks
+
+
+def get_default_config(module: nn.Module, device_ids: Sequence[torch.device]) -> Config:
```
Comment (BlackSamorez) [inline src/tensor_parallel/autoconfig.py:12]: It's now a function in a separate file to prevent circular imports

**Comment by BlackSamorez** [comment]
I've done a major refactoring around configs, wrappers.
The new dependency graph:
The new dependency graph

Notably:
* `Config` doesn't know anything about `TensorParallelWrapper`
* `TensorParallelWrapper` doesn't know anything about model shards.

**Comment by BlackSamorez** [comment]
Moved autoconfig under the hood.
The latest dependency graph:
tensor_parallel

File: src/tensor_parallel/state_actions.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
+"""slicing_actions.py: file contains actions """
+
+from abc import ABC, abstractclassmethod
+from typing import Any, Sequence
+
+import torch
+from torch import Tensor
+
+
+class StateAction(ABC):
```
Comment (BlackSamorez) [inline src/tensor_parallel/state_actions.py:10]: What I did with `state_rules`:
* They are now `Dict[str, StateAction]` where `StateAction` is an interface.
* `StateAction` has two methods:
*  * `def __call__(self, tensor: Tensor, rank: int) -> Tensor:` - split the tensor and return the corresponding part.
*  * `def undo(self, tensors: Sequence[Tensor]) -> Tensor:` - merge the parts back for initial tensor reconstruction.

**Review by justheuristic — APPROVED** [review]
To the best of my understanding, this is a major breaking change. If/when you release it, i would recommend updating all integrations and examples -- AND incrementing the major version, e.g. 2.x

**Comment by BlackSamorez** [comment]
> To the best of my understanding, this is a major breaking change. If/when you release it, i would recommend updating all integrations and examples -- AND incrementing the major version, e.g. 2.x

Maybe I should add backwards compatibility for configs.
Something in the line of creating a custom `StateAction` if it's a callable or a tuple of callables and throwing a deprecation warning.

**Comment by BlackSamorez** [comment]
> > To the best of my understanding, this is a major breaking change. If/when you release it, i would recommend updating all integrations and examples -- AND incrementing the major version, e.g. 2.x
> 
> Maybe I should add backwards compatibility for configs. Something in the line of creating a custom `StateAction` if it's a callable or a tuple of callables and throwing a deprecation warning.

Done

**Pull request merged**